components/wifi-controller/espidf-v4.0.c

```diff
@@ -0,0 +1,176 @@
+    }
+    else if (event_base == IP_EVENT && event_id == IP_EVENT_STA_GOT_IP)
+    {
+        ESP_LOGD(WIFI_LOG, "got IP");
+        xEventGroupSetBits(s_wifi_event_group, WIFI_CONNECTED_BIT);
+        heap_caps_check_integrity_all(true);
+    }
+}
+
+esp_err_t wifi_controller_connect(wifi_config_t wifi_config)
+{
+    if (!initialized)
+    {
+        wifi_controller_init();
+    }
+
+    ESP_ERROR_CHECK(esp_event_handler_register(WIFI_EVENT, ESP_EVENT_ANY_ID, &event_handler, NULL));
+    ESP_ERROR_CHECK(esp_event_handler_register(IP_EVENT, IP_EVENT_STA_GOT_IP, &event_handler, NULL));
+
+    ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
+    ESP_ERROR_CHECK(esp_wifi_set_config(ESP_IF_WIFI_STA, &wifi_config));
+    ESP_ERROR_CHECK(esp_wifi_start());
+
+    s_wifi_event_group = xEventGroupCreate();
+
+    EventBits_t bits = xEventGroupWaitBits(s_wifi_event_group, WIFI_CONNECTED_BIT | WIFI_FAIL_BIT, pdFALSE, pdFALSE, portMAX_DELAY);
+
+    ESP_ERROR_CHECK(esp_event_handler_unregister(WIFI_EVENT, ESP_EVENT_ANY_ID, event_handler));
+    vEventGroupDelete(s_wifi_event_group);
+
+    if (bits & WIFI_CONNECTED_BIT)
+    {
+        ESP_LOGV(WIFI_LOG, "connected to ap SSID:%s", wifi_config.sta.ssid);
+        return ESP_OK;
+    }
+    else if (bits & WIFI_FAIL_BIT)
+    {
+        ESP_LOGI(WIFI_LOG, "Failed to connect to SSID:%s with PASSWORD:%s", wifi_config.sta.ssid, wifi_config.sta.password);
+        return ESP_ERR_INVALID_ARG;
+    }
+    else
+    {
+        ESP_LOGE(WIFI_LOG, "UNEXPECTED EVENT");
+        return ESP_FAIL;
+    }
+}
+
+esp_err_t wifi_controller_reconnect(void)
+{
+    if (!initialized)
+    {
+        wifi_controller_init();
+    }
+
+    ESP_ERROR_CHECK(esp_event_handler_register(WIFI_EVENT, ESP_EVENT_ANY_ID, &event_handler, NULL));
+    ESP_ERROR_CHECK(esp_event_handler_register(IP_EVENT, IP_EVENT_STA_GOT_IP, &event_handler, NULL));
+
+    ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
+    ESP_ERROR_CHECK(esp_wifi_start());
+
+    s_wifi_event_group = xEventGroupCreate();
+
+    EventBits_t bits = xEventGroupWaitBits(s_wifi_event_group, WIFI_CONNECTED_BIT | WIFI_FAIL_BIT, pdFALSE, pdFALSE, portMAX_DELAY);
+
+    ESP_ERROR_CHECK(esp_event_handler_unregister(WIFI_EVENT, ESP_EVENT_ANY_ID, event_handler));
+    vEventGroupDelete(s_wifi_event_group);
+
+    if (bits & WIFI_CONNECTED_BIT)
+    {
+        ESP_LOGV(WIFI_LOG, "reconnected to last ap");
+        return ESP_OK;
+    }
+    else if (bits & WIFI_FAIL_BIT)
+    {
+        ESP_LOGI(WIFI_LOG, "Failed to reconnect to last ap");
+        return ESP_ERR_INVALID_ARG;
+    }
+    else
+    {
+        ESP_LOGE(WIFI_LOG, "UNEXPECTED EVENT");
+        return ESP_FAIL;
+    }
+}
+
+wifi_ap_record_t *wifi_controller_connection(void)
+{
+    if (esp_wifi_sta_get_ap_info(&current_wifi_ap) == ESP_OK)
+    {
+        ESP_LOGD(WIFI_LOG, "Current AP: %s", current_wifi_ap.ssid);
+        return &current_wifi_ap;
+    }
+    return NULL;
+}
```

Comment:
Please compare this files contents with espidf-v4.1.c, and confirm that there are no unintentional changes.
Intention was to build a version for API version 4.0.